plugins/web/opentelemetry-plugin-react-load/src/BaseOpenTelemetryComponent.ts

```diff
@@ -0,0 +1,541 @@
+   */
+  private _instrumentFunction(
+    react: React.Component,
+    spanName: string,
+    original: any
+  ) {
+    const span = this._createSpanWithParent(react, spanName);
+    let res;
+    try {
+      res = original();
+    } catch (err) {
+      if (span) {
+        span.setAttribute(AttributeNames.REACT_ERROR, err.stack);
+        span.end();
+      }
+      this._endParentSpan(react);
+      throw err;
+    }
+
+    if (span) {
+      span.end();
+    }
+    return res;
+  }
+
+  /**
+   * Ends the current parent span.
+   * @param react React component parent span belongs to.
+   */
+  private _endParentSpan(react: React.Component) {
+    const parentSpanData = this._parentSpanMap.get(react);
+    if (parentSpanData) {
+      parentSpanData.span.updateName(parentSpanData.name);
+      parentSpanData.span.end();
+      this._parentSpanMap.delete(react);
+    }
+  }
+
+  /**
+   * Returns attributes object for spans
+   * @param react React component currently being instrumented
+   **/
+  private _getAttributes(react: React.Component) {
+    let state = '';
+    try {
+      state = JSON.stringify(react.state);
+    } catch {
+      state = '{"message": "state could not be turned into string"}';
+    }
+    return {
+      [GeneralAttribute.COMPONENT]: this.moduleName,
+      [AttributeNames.LOCATION_URL]: window.location.href,
+      [AttributeNames.REACT_NAME]: react.constructor.name,
+      [AttributeNames.REACT_STATE]: state,
+    };
+  }
+
+  /**
+   * Returns the current parent span. If one doesn't exist,
+   * this function creates a parent span
+   * @param react React component parent span belongs to.
+   */
+  private _getCreateParentSpan(
+    react: React.Component
+  ): ParentSpanData | undefined {
+    const parentSpan: ParentSpanData | undefined = this._parentSpanMap.get(
+      react
+    );
+    if (!parentSpan) {
+      const span = this._createSpan(react, DEFAULT_PARENT_NAME);
+      if (span) {
+        this._parentSpanMap.set(react, { span, name: DEFAULT_PARENT_NAME });
+      }
+    }
+    return this._parentSpanMap.get(react);
+  }
+
+  /**
+   * Patches the render lifecycle method
+   */
+  private _patchRender() {
+    return (original: RenderFunction): RenderFunction => {
+      const plugin = this;
+      return function patchRender(
+        this: React.Component,
+        ...args
+      ): React.ReactNode {
+        // Render is the first method in the mounting flow, if parent span wasn't named updating already then we're in the mounting flow
+        // The naming assigned here will only be reflected if the component throws and error when rendering
+        const parentData = plugin._getCreateParentSpan(this);
+        if (parentData && parentData.name === DEFAULT_PARENT_NAME) {
+          parentData.name = AttributeNames.MOUNTING_SPAN;
+        }
+        return plugin._instrumentFunction(this, 'render', () => {
+          return original!.apply(this, args);
```

Comment:
Related to the comment below! The type of the original function can be undefined as specified by React, but in my code I guarantee it to be a no-op when undefined. This patch function has to have the same signature as the actual React method so that the shimmer wrap function can take them as arguments